File: sdk/src/trading_sdk/wallet.py
from typing_extensions import Protocol, TypeVar, TypedDict, NotRequired
from decimal import Decimal

from trading_sdk.types import Num
from trading_sdk.errors import AuthedError

S = TypeVar('S', bound=str)

class Wallet(Protocol):
  async def withdraw(
    self, currency: str, *, address: str, amount: Num,
    network: str | None = None,
  ) -> AuthedError | None:
    ...

  async def get_deposit_address(self, currency: str, *, network: str | None = None) -> str | AuthedError:
    ...

  class WithdrawalMethod(TypedDict):
    network: str
    fee: Decimal
    min_amount: NotRequired[Decimal]

  async def get_withdrawal_methods(self, *currencies: S) -> dict[S, WithdrawalMethod] | AuthedError:
    ...

File: rust/src/wallet.rs
use crate::errors::AuthedError;

pub struct WithdrawalMethod {
  pub network: String,
  pub fee: String,
  pub min_amount: Option<String>,
}

#[async_trait::async_trait]
pub trait Wallet {
  async fn withdraw(&self, currency: &str, address: &str, amount: String, network: Option<&str>) -> Result<(), AuthedError>;
  async fn get_deposit_address(&self, currency: &str, network: Option<&str>) -> Result<String, AuthedError>;
  async fn get_withdrawal_methods(&self, currency: &str) -> Result<Vec<WithdrawalMethod>, AuthedError>;
} 